x/cronos/keeper/precompiles/ica.go

```diff
@@ -0,0 +1,199 @@
+	)
+}
+
+type IcaContract struct {
+	cdc                 codec.Codec
+	channelKeeper       *ibcchannelkeeper.Keeper
+	icaControllerKeeper *icacontrollerkeeper.Keeper
+}
+
+func NewIcaContract(
+	cdc codec.Codec,
+	channelKeeper *ibcchannelkeeper.Keeper,
+	icaControllerKeeper *icacontrollerkeeper.Keeper,
+) precompiles.StatefulPrecompiledContract {
+	return &IcaContract{
+		cdc,
+		channelKeeper,
+		icaControllerKeeper,
+	}
+}
+
+func (ic *IcaContract) Address() common.Address {
+	return IcaContractAddress
+}
+
+// RequiredGas calculates the contract gas use
+func (ic *IcaContract) RequiredGas(input []byte) uint64 {
+	return RelayerContractRequiredGas
+}
+
+func (ic *IcaContract) IsStateful() bool {
+	return true
+}
+
+func (ic *IcaContract) Run(evm *vm.EVM, contract *vm.Contract, readonly bool) ([]byte, error) {
+	// parse input
+	methodID := contract.Input[:4]
+	stateDB := evm.StateDB.(precompiles.ExtStateDB)
+	var err error
+	var res codec.ProtoMarshaler
+	switch string(methodID) {
+	case string(RegisterAccountMethod.ID):
+		if readonly {
+			return nil, errors.New("the method is not readonly")
+		}
+		args, err := RegisterAccountMethod.Inputs.Unpack(contract.Input[4:])
+		if err != nil {
+			return nil, errors.New("fail to unpack input arguments")
+		}
+		connectionID := args[0].(string)
+		account := args[1].(common.Address)
+		version := args[2].(string)
+		txSender := evm.Origin
+		if !isSameAddress(account, contract.CallerAddress) && !isSameAddress(account, txSender) {
+			return nil, errors.New("unauthorized account registration")
+		}
+		err = stateDB.ExecuteNativeAction(func(ctx sdk.Context) error {
+			res, err = ic.icaControllerKeeper.RegisterInterchainAccount(ctx, &icacontrollertypes.MsgRegisterInterchainAccount{
+				Owner:        sdk.AccAddress(account.Bytes()).String(),
+				ConnectionId: connectionID,
+				Version:      version,
+			})
+			return err
+		})
+	case string(QueryAccountMethod.ID):
+		args, err := QueryAccountMethod.Inputs.Unpack(contract.Input[4:])
+		if err != nil {
+			return nil, errors.New("fail to unpack input arguments")
+		}
+		connectionID := args[0].(string)
+		account := args[1].(common.Address)
+		err = stateDB.ExecuteNativeAction(func(ctx sdk.Context) error {
+			res, err = ic.icaControllerKeeper.InterchainAccount(ctx, &icacontrollertypes.QueryInterchainAccountRequest{
+				Owner:        sdk.AccAddress(account.Bytes()).String(),
+				ConnectionId: connectionID,
+			})
+			return err
+		})
+	case string(SubmitMsgsMethod.ID):
+		if readonly {
+			return nil, errors.New("the method is not readonly")
+		}
+		args, err := SubmitMsgsMethod.Inputs.Unpack(contract.Input[4:])
+		if err != nil {
+			return nil, errors.New("fail to unpack input arguments")
+		}
+		connectionID := args[0].(string)
+		account := args[1].(common.Address)
+		data := args[2].(string)
+		timeout := args[3].(*big.Int)
+		txSender := evm.Origin
+		if !isSameAddress(account, contract.CallerAddress) && !isSameAddress(account, txSender) {
```

Comment:
Is `isSameAddress(account, contract.CallerAddress)` enough?

If we enforce owner == tx.origin then we can never call ica from contract address

Reply:
need some [change](https://github.com/cosmos/ibc-go/compare/v7.2.0...mmsqe:ibc-go:v7.2.0-rc2?expand=1) in ibc to handle generic

Reply:
not sure the enforce you mean, but current condition is allow call from CallerAddress or txSender

Reply:
I see, I was concerned that allowing from `tx.origin` may have some security issues but I think should be fine